### PR TITLE
#12078: SDPA generates its own attention mask

### PR DESCRIPTION
**Pull request by cglagovichTT**
### Ticket
#12078 

### Problem description
For long sequence lengths, the causal attention mask can grow to be multiple gigabytes. This blocks long context prefill in LLMs. 

### What's changed
This change modifies the SDPA kernel to make it create its own causal attention mask.

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/10739573109
- [x] T3k frequent tests: https://github.com/tenstorrent/tt-metal/actions/runs/10792151514
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by caixunshiren — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/transformer/sdpa/device/sdpa_op.cpp [hunk 1]
```diff
@@ -21,24 +21,29 @@ void ScaledDotProductAttention::validate(
...
-    TT_FATAL(input_tensors.at(0).device() == mask.device());
-    TT_FATAL(mask.get_layout() == Layout::TILE);
-    TT_FATAL(mask.get_dtype() == DataType::BFLOAT16 || mask.get_dtype() == DataType::BFLOAT8_B);
+    const auto& mask_option = optional_input_tensors.at(0);
+    if (mask_option.has_value()){
+        auto mask = optional_input_tensors.at(0).value();
+        TT_FATAL(mask.storage_type() == StorageType::DEVICE, "Operands to SDPA need to be on device!");
+        TT_FATAL(input_tensors.at(0).device() == mask.device());
+        TT_FATAL(mask.get_layout() == Layout::TILE);
+        TT_FATAL(mask.get_dtype() == DataType::BFLOAT16 || mask.get_dtype() == DataType::BFLOAT8_B);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/transformer/sdpa/device/sdpa_op.cpp:30]: These error messages should say something like, "When mask is provided to SPDA ... [expectation] [exact value provided which is not aligned with expectations]"

File: ttnn/cpp/ttnn/operations/transformer/sdpa/device/sdpa_op.cpp [hunk 2]
```diff
@@ -21,24 +21,29 @@ void ScaledDotProductAttention::validate(
...
 
-    TT_FATAL(mask.buffer()->buffer_type() == tt::tt_metal::BufferType::DRAM);
+        TT_FATAL(mask.buffer()->buffer_type() == tt::tt_metal::BufferType::DRAM);
+    }
 
     const auto q_shape = input_tensors.at(0).get_legacy_shape();
     const auto k_shape = input_tensors.at(1).get_legacy_shape();
     const auto v_shape = input_tensors.at(2).get_legacy_shape();
-    const auto mask_shape = mask.get_legacy_shape();
+    // const auto mask_shape = mask.get_legacy_shape();
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/transformer/sdpa/device/sdpa_op.cpp:38]: remove?

File: ttnn/cpp/ttnn/operations/transformer/sdpa/device/sdpa_op.cpp [hunk 3]
```diff
@@ -62,11 +66,18 @@ void ScaledDotProductAttention::validate(
         // Check qkv heads
         TT_FATAL(q_shape[-3] >= k_shape[-3]);
 
-        TT_FATAL(mask_shape[-3] == 1);
+        if (mask_option.has_value()){
+            const auto mask_shape = mask_option.value().get_legacy_shape();
+            TT_FATAL(q_shape[-4] == mask_shape[-4]);
+            TT_FATAL(q_shape[-2] == mask_shape[-2] && q_shape[-2] == mask_shape[-1]);
+            TT_FATAL(mask_shape[-3] == 1);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/transformer/sdpa/device/sdpa_op.cpp:73]: Same as above. Python users who will see error message generated by current code will find it useless.

File: ttnn/cpp/ttnn/operations/transformer/sdpa/device/sdpa_program_factory.cpp [hunk 4]
```diff
@@ -107,7 +107,7 @@ operation::ProgramWithCallbacks sdpa_multi_core(
     auto k_buffer = input_tensor_k.buffer();
     auto v_buffer = input_tensor_v.buffer();
     auto mask_buffer = attn_mask.has_value() ? attn_mask.value().buffer() : nullptr;
-    TT_FATAL(mask_buffer != nullptr);
+    // TT_FATAL(mask_buffer != nullptr);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/transformer/sdpa/device/sdpa_program_factory.cpp:110]: remove

**Review by ayerofieiev-tt — APPROVED** [review]

**Review by cfjchu — APPROVED** [review]

**Review by yieldthought — APPROVED** [review]

**Review by johanna-rock-tt — APPROVED** [review]
lgtm!

**Pull request merged**